rpc/internal/proc/proc.go

```diff
@@ -0,0 +1,115 @@
+// Package proc assits managing the lifecycle of goroutines.
+package proc
+
+import (
+	"context"
+	"sync"
+)
+
+// Spawn a new process.
+//
+// A process can be in one of the following states:
+//
+// - Running
+// - Stopping
+// - Stopped
+//
+// ...starting in the Running state.
+//
+// f is a callback that implements the actual logic of the process. It will be
+// passed a child context of ctx, and a Self. It should:
+//
+// 1. Run until its context is canceled is canceled (or its work is done).
+// 2. invoke BeginShutdown() on the Self that was passed to it, which
+//    transitions the process to the Stopping state.
+// 3. Perform any shutdown logic that is needed.
+// 4. Return. This transitions the process into the Stopped state.
+//
+// The state argument will be accessible by clients of the process via WithLive
+// until the process enters the Stopping state, after which the state is owned
+// by the process (returned by Self.BeginShutdown)
+//
+// Returns a handle, which can be used by clients of the process to interact
+// with its lifecycle.
+func Spawn[S any](ctx context.Context, state S, f func(context.Context, Self[S])) Handle[S] {
+	ctx, cancel := context.WithCancel(ctx)
+	proc := &proc[S]{
+		cancel: cancel,
+		done:   make(chan struct{}),
+		state:  state,
+	}
+	handle := Handle[S]{proc}
+	self := Self[S]{proc}
+	go func() {
+		defer func() {
+			self.BeginShutdown()
+			close(proc.done)
```

Comment:
I think these should probably be two separate `defer` statements, because we want to make sure `close(proc.done)` happens despite any panics in `f()`.

Reply:
Done.